### PR TITLE
Feature/export

### PR DESCRIPTION
**Pull request by fedelopez77**
* Added features to export embeddings
 * Keen and movie logic in separate files
 * Filtering of keen dataset based on min and max interactions of keens with users and viceversa
 * Computation of distance matrix in preprocessing to make training faster
 * Exporting metrics to file from different parallel runs. Script to process that file as well

### CODE REVIEW

File: hyperbolic-rs/rudders/models/euclidean.py [hunk 1]
```diff
@@ -37,14 +37,18 @@ class SMFactor(BaseEuclidean):
     """Simple Matrix Factorization model."""
 
     def score(self, user_embeds, item_embeds, all_pairs):
-        """Score based on dot product"""
+        """Score based on dot product (cosine similarity)"""
```
Comment (mscholz) [inline hyperbolic-rs/rudders/models/euclidean.py:40]: Thanks for pointing out the "cosine similarity" part!
For hyperbolic embeddings, the norm plays a crucial role, so a baseline that does not remove the norm aspect from these Euclidean embeddings might be interesting to look at as well?

File: hyperbolic-rs/export.py [hunk 2]
```diff
@@ -23,12 +25,16 @@
 def main():
```
[the remark this answers is not in the document]
Comment (fedelopez77) [reply hyperbolic-rs/export.py:25]: Done

File: hyperbolic-rs/export.py [hunk 3]
```diff
@@ -37,27 +43,59 @@ def main():
...
+
 
+def export_for_projector(filename, user_embeds, item_embeds, id2title, samples, closest_user_item, closest_item_item):
+    meta, coords = ["type\ttitle\tinteractions\tclosest"], []
+    for i, embed in enumerate(user_embeds):
+        coords.append("\t".join([str(x) for x in embed]))
 
-def to_hyperbolic(embeds):
-    return expmap0(tf.convert_to_tensor(embeds), tf.convert_to_tensor([1.0], dtype=tf.float64)).numpy()
+        interactions = [id2title.get(item_id, "None").replace("\t", "") for item_id in samples[i]]
```
[the remark this answers is not in the document]
Comment (fedelopez77) [reply hyperbolic-rs/export.py:83]: Done

File: hyperbolic-rs/export.py [hunk 4]
```diff
@@ -110,10 +144,26 @@ def plot(filename, user_embeds, item_embeds, subsample=0.5, scale=False, hyperbo
...
+    user_embeds = concat(x[:split], y[:split])
+    item_embeds = concat(x[split:], y[split:])
+    return user_embeds, item_embeds
+
+
+def plot(filename, user_embeds, item_embeds, subsample=0.5, alpha=0.75, size=2):
+    """
+    :param filename: to store the image
+    :param user_embeds: Numpy array of shape (len(user_embeds), 2)
+    :param item_embeds: Numpy array of shape (len(item_embeds), 2)
```
[the remark this answers is not in the document]
Comment (fedelopez77) [reply hyperbolic-rs/export.py:158]: Done

File: hyperbolic-rs/preprocess.py [hunk 5]
```diff
@@ -108,18 +112,75 @@ def plot_graph(samples):
...
+        distance_matrix = np.ones((len(iid2id), len(iid2id))) * -1
+
+    for src_iid, src_index in iid2id.items():
+        if src_iid not in item_item_distances_dict:
+            continue
+        src_dist = item_item_distances_dict[src_iid]
+        for dst_iid, distance in src_dist.items():
+            if src_iid != dst_iid and dst_iid in iid2id:
+                dst_index = iid2id[dst_iid]
+                distance_matrix[src_index, dst_index] = max(distance, 0.1)  # set a minimum distance for stability
```
[the remark this answers is not in the document]
Comment (fedelopez77) [reply hyperbolic-rs/preprocess.py:150]: Done

File: hyperbolic-rs/rudders/models/euclidean.py (hunk 1 above)
Comment (fedelopez77) [reply hyperbolic-rs/rudders/models/euclidean.py:40]: I'll try this out in the new set of models/experiments

**Pull request merged**